### PR TITLE
bootloader: extend managed assets bootloader interface to compose a candidate command line

### PR DESCRIPTION
**Pull request by bboozzoo**
Where the candidate command line is based on the current internal edition of the
asset.

### CODE REVIEW

File: boot/cmdline.go [hunk 1]
```diff
@@ -134,50 +134,64 @@ func getBootloaderManagingItsAssets(where string, opts *bootloader.Options) (boo
...
-	return mbl.CommandLine("snapd_recovery_mode=recover", systemArg, extraArgs)
+	if currentOrCandidate == currentEdition {
+		return mbl.CommandLine(modeArg, systemArg, extraArgs)
+	} else {
+		return mbl.CandidateCommandLine(modeArg, systemArg, extraArgs)
+	}
+}
+
+// ComposeRecoveryCommandLine composes the kernel command line used when booting
+// a given recovery mode system.
```
Comment (cmatsuoka) [inline boot/cmdline.go:181]: Maybe we could rephrase this as "booting a given system in recover mode"?

File: boot/cmdline.go [hunk 2]
```diff
@@ -134,50 +134,64 @@ func getBootloaderManagingItsAssets(where string, opts *bootloader.Options) (boo
...
-func ComposeRecoveryCommandLine(model *asserts.Model, system string) (string, error) {
+const (
+	currentEdition = iota
+	candidateEdition
+
+	runBootloader = iota
+	recoveryBootloader
+)
+
+func composeCommandLine(model *asserts.Model, currentOrCandidate, runOrRecovery int, system string) (string, error) {
```
Comment (cmatsuoka) [inline boot/cmdline.go:145]: The mode name is "recover" so I believe `runOrRecovery` should be `runOrRecover`.

Comment (bboozzoo) [reply boot/cmdline.go:145]: hm I guess I should `s/recoveryBootloader/recoverBootloader` too

Comment (bboozzoo) [reply boot/cmdline.go]: Actually we already have `ModeRun`, `ModeRecover`, so I tweaked the code to pass those identifiers instead.

File: boot/cmdline.go [hunk 3]
```diff
@@ -134,50 +134,64 @@ func getBootloaderManagingItsAssets(where string, opts *bootloader.Options) (boo
...
 		if err == errBootConfigNotManaged {
 			return "", nil
 		}
 		return "", err
 	}
-	systemArg := fmt.Sprintf("snapd_recovery_system=%v", system)
 	// TODO:UC20: fetch extra args from gadget
 	extraArgs := ""
-	return mbl.CommandLine("snapd_recovery_mode=recover", systemArg, extraArgs)
+	if currentOrCandidate == currentEdition {
```
Comment (cmatsuoka) [inline boot/cmdline.go:173]: When we test whether to retrieve the current or the candidate command line we may have already failed when we get the bootloader (in line 164 above). This would be the same situation that caused the previous error during install.

File: boot/cmdline.go (hunk 2 above)
Comment (pedronis) [reply boot/cmdline.go]: to be clear, RecoveryBootloader is a fine name.  is just the actual mode values that are verbs

File: boot/cmdline.go (hunk 3 above)
Comment (bboozzoo) [reply boot/cmdline.go:166]: Per discussion with @pedronis we will not check whether the on disk asset is managed. The *CommandLine() call is part of the interface of a managed-assets bootloader, and the boot does the call only after checking then model, which when right implies we always have a managed asset installed (otherwise we cannot fulfill what model promises). I've added a fallback for ComandLine() which assumes a well known edition (1) when the managed asset is not installed, since it matches the base edition of grub.cfg we imported into the tree.

File: boot/cmdline.go [hunk 4]
```diff
@@ -124,60 +124,67 @@ func getBootloaderManagingItsAssets(where string, opts *bootloader.Options) (boo
...
+
+func composeCommandLine(model *asserts.Model, currentOrCandidate int, mode, system string) (string, error) {
 	if model.Grade() == asserts.ModelGradeUnset {
 		return "", nil
 	}
-
+	if mode != ModeRun && mode != ModeRecover {
+		return "", fmt.Errorf("internal error: unsupported command line mode %q", mode)
+	}
 	// get the ubuntu-seed bootloader
```
Comment (pedronis) [inline boot/cmdline.go:142]: this is comment is a bit confusing, we will get either the boot one and the seed one in the if below

File: boot/cmdline.go [hunk 5]
```diff
@@ -124,60 +124,67 @@ func getBootloaderManagingItsAssets(where string, opts *bootloader.Options) (boo
...
-	// TODO:UC20: fetch extra args from gadget
-	extraArgs := ""
-	return mbl.CommandLine("snapd_recovery_mode=run", "", extraArgs)
+	return composeCommandLine(model, currentEdition, ModeRun, "")
+}
+
+// ComposeCandidateCommandLine composes the kernel command line used when
+// booting the system in run mode with the current built-in edition of managed
+// boot assets.
+func ComposeCandidateCommandLine(model *asserts.Model) (string, error) {
```
Comment (pedronis) [inline boot/cmdline.go:188]: we'll need ComposeCandidateRecoveryCommandLine at some point, not immediately but at least maybe have a placeholder TODO

**Comment by bboozzoo** [comment]
The following fail due to troubles downloading the rootfs images from `ubuntu` (cloud-images.ubuntu.com/) remote of lxd:
- google:ubuntu-16.04-64:tests/main/lxd
- google:ubuntu-18.04-64:tests/main/lxd
- google:ubuntu-20.04-64:tests/main/lxd

@pedronis can you use your superpowers to merge the PR?

File: bootloader/grub.go [hunk 6]
```diff
@@ -421,6 +408,46 @@ func (g *grub) CommandLine(modeArg, systemArg, extraArgs string) (string, error)
...
+	if err != nil {
+		if err != errNoEdition {
+			return "", fmt.Errorf("cannot obtain edition number of current boot config: %v", err)
+		}
+		// we were called using the ManagedAssetsBootloader interface
+		// meaning the caller expects to us to use the managed assets,
+		// since one on disk is not managed, use the initial edition of
+		// the internal boot asset which is compatible with grub.cfg
+		// used before we started writing out the files ourselves
+		edition = 1
```
Comment (anonymouse64) [inline bootloader/grub.go:430]: ahhh I see the issue now, makes sense to me, thanks for this comment
